homeassistant/components/pvpc_hourly_pricing/sensor.py

```diff
@@ -0,0 +1,394 @@
+        attributes["next_best_at"] = list(
+            map(
+                lambda x: _local(x).hour,
+                filter(lambda x: x >= utc_time, prices_sorted.keys()),
+            )
+        )
+        for ts_utc, price_h in self._current_prices.items():
+            ts_local = _local(ts_utc)
+            if ts_local.day > actual_time.day:
+                attr_key = f"price_next_day_{ts_local.hour:02d}h"
+            elif ts_local.day < actual_time.day:
+                attr_key = f"price_last_day_{ts_local.hour:02d}h"
+            else:
+                attr_key = f"price_{ts_local.hour:02d}h"
+            if attr_key in attributes:  # DST change with duplicated hour :)
+                attr_key += "_d"
+            attributes[attr_key] = price_h
+
+        self._attributes = attributes
+        return True
+
+    async def async_update(self, *_args):
+        """Update the sensor state."""
+        if not self._init_done:
+            # abort until added_to_hass is finished
+            return
+
+        if self._process_state_and_attributes():
+            await self.async_update_ha_state()
+        else:
+            # If no prices present, download and schedule a future state update
+            self._state_available = False
+            self.async_schedule_update_ha_state()
+
+            if self._data_source_available:
+                _LOGGER.debug(
+                    "[%s]: Downloading prices as there are no valid ones",
+                    self.entity_id,
+                )
+                async_track_point_in_time(
+                    self.hass,
+                    self.async_update,
+                    dt_util.now() + timedelta(seconds=self._timeout),
+                )
+            await self.async_update_prices()
+
+    async def _download_official_data(self, day: date) -> List[dict]:
+        """Make GET request to 'api.esios.ree.es' to retrieve hourly prices."""
+        url = _RESOURCE.format(day=day)
+        try:
+            with async_timeout.timeout(self._timeout, loop=self.hass.loop):
+                resp = await self._websession.get(url)
+                if resp.status < 400:
+                    data = await resp.json()
+                    return data["PVPC"]
+        except KeyError:
+            _LOGGER.debug("Bad try on getting prices for %s", day)
+        except asyncio.TimeoutError:  # pragma: no cover
+            if self._data_source_available:
+                _LOGGER.warning("Timeout error requesting data from '%s'", url)
+        except aiohttp.ClientError:  # pragma: no cover
+            if self._data_source_available:
+                _LOGGER.warning("Client error in '%s'", url)
+        return {}
+
+    async def async_update_prices(self, *_args):
+        """Update electricity prices from the ESIOS API."""
+        localized_now = dt_util.utcnow().astimezone(_REFERENCE_TZ)
+        data = await self._download_official_data(localized_now.date())
```

Comment:
Please move this processing into the pypi package

Reply:
Done